src/modules/multi-chain-share/DestinationMinter.sol

```diff
@@ -0,0 +1,157 @@
+
+    //============================== ERRORS ===============================
+
+    error DestinationMinter___SourceChainNotAllowlisted(uint64 sourceChainSelector);
+    error DestinationMinter___SenderNotAllowlisted(address sender);
+    error DestinationMinter___InvalidTo();
+    error DestinationMinter___FeeTooHigh();
+
+    //============================== EVENTS ===============================
+
+    event BridgeToSource(uint256 amount, address to);
+    event BridgeFromSource(uint256 amount, address to);
+
+    //============================== MODIFIERS ===============================
+
+    modifier onlyAllowlisted(uint64 _sourceChainSelector, address _sender) {
+        if (_sourceChainSelector != sourceChainSelector)
+            revert DestinationMinter___SourceChainNotAllowlisted(_sourceChainSelector);
+        if (_sender != targetSource) revert DestinationMinter___SenderNotAllowlisted(_sender);
+        _;
+    }
+
+    //============================== IMMUTABLES ===============================
+
+    /**
+     * @notice The address of the SourceLocker on source chain.
+     */
+    address public immutable targetSource;
+
+    /**
+     * @notice The CCIP source chain selector.
+     */
+    uint64 public immutable sourceChainSelector;
+
+    /**
+     * @notice The CCIP destination chain selector.
+     */
+    uint64 public immutable destinationChainSelector;
+
+    /**
+     * @notice This networks LINK contract.
+     */
+    ERC20 public immutable LINK;
+
+    /**
+     * @notice The message gas limit to use for CCIP messages.
+     */
+    uint256 public immutable messageGasLimit;
+
+    constructor(
+        address _router,
+        address _targetSource,
+        string memory _name,
+        string memory _symbol,
+        uint8 _decimals,
+        uint64 _sourceChainSelector,
+        uint64 _destinationChainSelector,
+        address _link,
+        uint256 _messageGasLimit
+    ) ERC20(_name, _symbol, _decimals) CCIPReceiver(_router) {
+        targetSource = _targetSource;
+        sourceChainSelector = _sourceChainSelector;
+        destinationChainSelector = _destinationChainSelector;
+        LINK = ERC20(_link);
+        messageGasLimit = _messageGasLimit;
+    }
+
+    //============================== BRIDGE ===============================
+
+    /**
+     * @notice Bridge shares back to source chain.
+     */
+    function bridgeToSource(uint256 amount, address to, uint256 maxLinkToPay) external returns (bytes32 messageId) {
+        if (to == address(0)) revert DestinationMinter___InvalidTo();
+        _burn(msg.sender, amount);
+
+        Client.EVM2AnyMessage memory message = _buildMessage(amount, to);
+
+        IRouterClient router = IRouterClient(this.getRouter());
+
+        uint256 fees = router.getFee(sourceChainSelector, message);
+
+        if (fees > maxLinkToPay) revert DestinationMinter___FeeTooHigh();
+
+        LINK.safeTransferFrom(msg.sender, address(this), fees);
```

Comment:
Hmmm that is a good question. The router address is verified to be safe, and these contracts will not be holding any customer funds that are LINK. So it would probably only help us if for some reason the router did not transfer the link it said it would, and the LINK contract had ERC20 breaking logic like USDT where we revert when going from non zero allowance to non zero allowance.